subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -60,24 +61,57 @@
 public class GroovyCompile extends AbstractCompile {
     private Compiler<GroovyJavaJointCompileSpec> compiler;
     private FileCollection groovyClasspath;
+    private FileCollection compilerPluginClasspath;
     private final CompileOptions compileOptions;
     private final GroovyCompileOptions groovyCompileOptions = new GroovyCompileOptions();
 
     public GroovyCompile() {
         CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
         this.compileOptions = compileOptions;
+        this.compilerPluginClasspath = getServices().get(ObjectFactory.class).fileCollection();
         CompilerForkUtils.doNotCacheIfForkingViaExecutable(compileOptions, getOutputs());
     }
 
+    @Override
+    @CompileClasspath
+    public FileCollection getClasspath() {
+        return super.getClasspath();
+    }
+
+    @Classpath
+    public FileCollection getCompilerPluginClasspath() {
+        return compilerPluginClasspath;
+    }
+
+    public void setCompilerPluginClasspath(FileCollection compilerPluginClasspath) {
+        if (!experimentalCompilationAvoidanceEnabled()) {
+            throw new IllegalStateException("You need to enable org.gradle.groovy.compilation.avoidance system property!");
+        }
+        this.compilerPluginClasspath = compilerPluginClasspath;
+    }
+
+    private boolean experimentalCompilationAvoidanceEnabled() {
+        return Boolean.getBoolean("org.gradle.groovy.compilation.avoidance");
+    }
+
     @Override
     @TaskAction
     protected void compile() {
         checkGroovyClasspathIsNonEmpty();
+        separateCompileClasspathIfCompileAvoidanceEnabled();
         DefaultGroovyJavaJointCompileSpec spec = createSpec();
         WorkResult result = getCompiler(spec).execute(spec);
         setDidWork(result.getDidWork());
     }
 
+    private void separateCompileClasspathIfCompileAvoidanceEnabled() {
```

Comment:
When compilation avoidance is disabled, we want new `GroovyCompile` behave exactly same as before. So we have to return `classpath` again in `compilerPluginClasspath` so that it can still be normalized as `@Classpath`.